File: python/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    # O(n) time | O(1) space
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float("inf")
        max_price = float("-inf")
        max_profit = 0
        for price in prices:
            if price < min_price:
                min_price = price
                max_price = price
            if price > max_price:
                max_price = price
                profit = max_price - min_price
                if profit > max_profit:
                    max_profit = profit
        return max_profit

File: c++/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    // O(n) time, O(1) space - where n is the size of the input array
    int maxProfit(vector<int>& prices) {
        int res = 0;
        int minPrice = numeric_limits<int>::max();
        int maxPrice = numeric_limits<int>::min();
        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price;
                maxPrice = price;
            }
            if (price > maxPrice) {
                maxPrice = price;
                int profit = maxPrice - minPrice;
                if (profit > res) {
                    res = profit;
                }
            }
        }
        return res;
    }
};